interactionlib.php

```diff
@@ -0,0 +1,193 @@
+     * @param string $action action string
+     * @return step_interactive_response defines if the step still wants to process this course
+     *      - proceed: the step has finished and respective controller class can take over.
+     *      - stillprocessing: the step still wants to process the course and is responsible for rendering the site.
+     *      - noaction: the action is not defined for the step.
+     *      - rollback: the step has finished and respective controller class should rollback the process.
+     */
+    public function handle_interaction($process, $step, $action = 'default') {
+        global $DB, $PAGE;
+        $PAGE->set_context(\context_system::instance());
+        $settings = settings_manager::get_settings($step->id, settings_type::STEP);
+        $type = 'student';
+        $typeid = 1;
+        if ($action == self::ACTION_NOTIFY) {
+            $settings[$type . '_content'] = format_text($settings[$type . '_content'], FORMAT_HTML);
+            $userstobeinformed = $DB->get_records('lifecyclestep_notifystudents',
+                array('courseid' => $process->courseid, 'instanceid' => $step->id, 'emailtype' => $typeid), '', 'distinct touser');
+            foreach ($userstobeinformed as $userrecord) {
+                $user = \core_user::get_user($userrecord->touser);
+                $transaction = $DB->start_delegated_transaction();
+                $mailentries = $DB->get_records('lifecyclestep_notifystudents',
+                    array('instanceid' => $step->id, 'courseid' => $process->courseid,
+                        'touser' => $user->id, 'emailtype' => $typeid));
+
+                $parsedsettings = (new notifystudents)->replace_placeholders($settings, $user, $step->id, $mailentries);
+
+                $subject = $parsedsettings[$type . '_subject'];
+                $contenthtml = $parsedsettings[$type . '_content'];
+                email_to_user($user, \core_user::get_noreply_user(), $subject, html_to_text($contenthtml), $contenthtml);
+                $DB->delete_records('lifecyclestep_notifystudents',
+                    array('instanceid' => $step->id, 'courseid' => $process->courseid,
+                        'touser' => $user->id, 'emailtype' => $typeid));
+                $transaction->allow_commit();
+            }
+        } else if ($action == self::ACTION_NONOTIFY) {
+            $DB->delete_records('lifecyclestep_notifystudents',
+                array('instanceid' => $step->id, 'courseid' => $process->courseid, 'emailtype' => $typeid));
+        }
+        return step_interactive_response::no_action();
+    }
+
+    /**
+     * Returns the due date.
+     * @param int $processid Id of the process.
+     * @param int $stepid Id of the step instance.
+     * @return string formatted date.
+     * @throws \coding_exception
+     * @throws \dml_exception
+     */
+    public function get_due_date($processid, $stepid) {
```

Comment:
I know that the email step does exactly the same, but I'm confused. Do you know why it checks every step, and not just the step specified by $stepid?

Reply:
I was also confused. I can test if there are any resulting problems when I remove it.

Reply:
This function is only used to display the due date in the action table, so it won't cause any 'big' problems, but in my understanding, this just returns the summated delay from this and any following notifystudent steps, and I don't know why the aggregated delay should be displayed in the table there.